url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/ssamot/mcts/backpropagators/EXP3Backpropagator.java

```diff
@@ -0,0 +1,48 @@
+/*
+ *
+ * *** BEGIN LICENSE
+ *  Copyright (C) 2012 Spyridon Samothrakis spyridon.samothrakis@gmail.com
+ *  This program is free software: you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 3, as published
+ *  by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful, but
+ *  WITHOUT ANY WARRANTY; without even the implied warranties of
+ *  MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * *** END LICENSE
+ *
+ */
+
+package ssamot.mcts.backpropagators;
+
+import java.util.List;
+
+import ssamot.mcts.selectors.exp3.EXP3Node;
+
+public class EXP3Backpropagator<T extends EXP3Node>implements Backpropagator<T> {
+
+	@Override
+	public void backpropagate(List<T> nodes, List<Double> reward) {
+		// TODO Auto-generated method stub
+		int nodeSize = nodes.size();
+
+		for (int i = nodeSize - 1; i >= 0; i--) {
+
+			T node = nodes.get(i);
+			int id = node.getRewardId();
+			if (node.getRewardId() != 0) {
+				double value = (reward.get(id));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;reward&#39; (index: 1 | type: List). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;nodes&#39;. The same parameter type is validated here for example: njr_failed/url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/lunar/LunarProblem.java:105. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).